File: model/src/main/java/io/github/runtimemodels/chazm/relation/UsesFactory.java
package io.github.runtimemodels.chazm.relation;

import io.github.runtimemodels.chazm.entity.Pmf;
import io.github.runtimemodels.chazm.entity.Role;
import io.github.runtimemodels.chazm.relation.Uses;

/**
 * The {@linkplain UsesFactory} interface defines the API for constructing {@linkplain Uses} relations.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
public interface UsesFactory {

	/**
	 * Constructs an {@linkplain Uses}.
	 *
	 * @param role
	 *            the {@linkplain Role} of the {@linkplain Uses}.
	 * @param pmf
	 *            the {@linkplain Pmf} of the {@linkplain Uses}.
	 * @return an {@linkplain Uses}.
	 */
	Uses buildUses(Role role, Pmf pmf);

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/factory/UsesFactory.kt
package runtimemodels.chazm.model.factory

import runtimemodels.chazm.api.entity.Pmf
import runtimemodels.chazm.api.entity.Role
import runtimemodels.chazm.api.relation.Uses

/**
 * The [UsesFactory] interface defines the API for constructing [Uses] relations.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
interface UsesFactory {
    /**
     * Constructs an [Uses].
     *
     * @param role the [Role] of the [Uses].
     * @param pmf  the [Pmf] of the [Uses].
     * @return an [Uses].
     */
    fun build(role: Role, pmf: Pmf): Uses
}
